include/seastar/core/semaphore.hh

```diff
@@ -482,6 +482,7 @@ public:
     semaphore_units(semaphore_units&& o) noexcept : _sem(o._sem), _n(std::exchange(o._n, 0)) {
     }
     semaphore_units& operator=(semaphore_units&& o) noexcept {
+        return_all();
         _sem = o._sem;
```

Comment:
This doesn't work when this == &o.

It technically doesn't need to work (passing a temporary contains a promise not to use it other than to invoke the destructor), but the damage is so great it's better to protect against it.

Reply:
fixed in v4